text/0000-name-resolution.md

```diff
@@ -0,0 +1,420 @@
+            process_names()
+            process_work_list()
+        }
+        expand_macros()
+    }
+
+    for item in work_list {
+        report_error()
+    } else {
+        success!()
+    }
+}
+
+fn process_names() {
+    // 'module' includes `mod`s, top level of the crate, function bodies
+    for each unseen item in any module {
+        if item is a definition {
+            // struct, trait, type, local variable def, etc.
+            bindings.insert(item.name, module, item)
+            populate_back_links(module, item)
+        } else {
+            try_to_resolve_import(module, item)
+        }
+        record_macro_uses()
+    }
+}
+
+fn try_to_resolve_import(module, item) {
+    if item is an explicit use {
+        // item is use a::b::c as d;
+        match try_to_resolve(item) {
+            Ok(r) => {
+                add(bindings.insert(d, module, r, Priority::Explicit))
+                populate_back_links(module, item)
+            }
+            Err() => work_list.push(module, item)
+        }
+    } else if item is a glob {
+        // use a::b::*;
+        match try_to_resolve(a::b) {
+            Ok(n) => 
+                for binding in n {
+                    bindings.insert_if_no_higher_priority_binding(binding.name, module, binding, Priority::Glob)
+                    populate_back_links(module, binding)
+                }
+                add_back_link(n to module)
+                work_list.remove()
+            Err(_) => work_list.push(module, item)
+        }
+    }    
+}
+
+fn process_work_list() {
+    for each (module, item) in work_list {
+        work_list.remove()
+        try_to_resolve_import(module, item)
+    }
+}
+```
+
+In order to keep macro expansion comprehensible to programmers, we must enforce
+that all macro uses resolve to the same binding at the end of resolution as they
+do when they were resolved.
+
+We rely on a monotonicity property in macro expansion - once an item exists in a
+certain place, it will always exist in that place. It will never disappear and
+never change. Note that for the purposes of this property, I do not consider
+code annotated with a macro to exist until it has been fully expanded.
+
+A consequence of this is that if the compiler resolves a name, then does some
+expansion and resolves it again, the first resolution will still be valid.
+However, another resolution may appear, so the resolution of a name may change
+as we expand. It can also change from a good resolution to an ambiguity. It is
+also possible to change from good to ambiguous to good again. There is even an
+edge case where we go from good to ambiguous to the same good resolution (but
+via a different route).
+
+If import resolution succeeds, then we check our record of name resolutions. We
+re-resolve and check we get the same result. We can also check for un-used
+macros at this point.
+
+### Privacy
+
+In order to resolve imports (and in the future for macro privacy), we must be
```

Comment:
No, there are no user-visible changes to privacy or macros in this RFC. I do plan to propose this later though.